docs/src/literate-tutorials/gpu_qp_heat_equation.jl

```diff
@@ -0,0 +1,182 @@
+function assemble_element_std!(Ke::Matrix, fe::Vector, cellvalues::CellValues)
+    n_basefuncs = getnbasefunctions(cellvalues)
+    ## Loop over quadrature points
+    for q_point in 1:getnquadpoints(cellvalues)
+        ## Get the quadrature weight
+        dΩ = getdetJdV(cellvalues, q_point)
+        ## Loop over test shape functions
+        for i in 1:n_basefuncs
+            δu = shape_value(cellvalues, q_point, i)
+            ∇δu = shape_gradient(cellvalues, q_point, i)
+            ## Add contribution to fe
+            fe[i] += δu * dΩ
+            ## Loop over trial shape functions
+            for j in 1:n_basefuncs
+                ∇u = shape_gradient(cellvalues, q_point, j)
+                ## Add contribution to Ke
+                Ke[i, j] += (∇δu ⋅ ∇u) * dΩ
+            end
+        end
+    end
+    return Ke, fe
+end
+
+
+function create_buffers(cellvalues, dh)
+    f = zeros(ndofs(dh))
+    K = allocate_matrix(dh)
+    assembler = start_assemble(K, f)
+    ## Local quantities
+    n_basefuncs = getnbasefunctions(cellvalues)
+    Ke = zeros(n_basefuncs, n_basefuncs)
+    fe = zeros(n_basefuncs)
+    return (; f, K, assembler, Ke, fe)
+end
+
+
+# Standard global assembly
+
+
+function assemble_global!(cellvalues, dh::DofHandler, qp_iter::Val{QPiter}) where {QPiter}
+    (; f, K, assembler, Ke, fe) = create_buffers(cellvalues, dh)
+    ## Loop over all cels
+    for cell in CellIterator(dh)
+        fill!(Ke, 0)
+        fill!(fe, 0)
+        if QPiter
+            ## reinit!(cellvalues, getcoordinates(cell))
+            assemble_element_qpiter!(Ke, fe, cellvalues, getcoordinates(cell))
+        else
+            ## Reinitialize cellvalues for this cell
+            reinit!(cellvalues, cell)
+            ## Compute element contribution
+            assemble_element_std!(Ke, fe, cellvalues)
+        end
+        ## Assemble Ke and fe into K and f
+        assemble!(assembler, celldofs(cell), Ke, fe)
+    end
+    return K, f
+end
+
+
+## gpu version of element assembly
+
+
+function assemble_element!(Ke, fe, cv, cell)
+    n_basefuncs = getnbasefunctions(cv)
+    for qv in Ferrite.QuadratureValuesIterator(cv, getcoordinates(cell))
+        ## Get the quadrature weight
+        dΩ = getdetJdV(qv)
+        ## Loop over test shape functions
+        for i in 1:n_basefuncs
+            δu = shape_value(qv, i)
+            ∇u = shape_gradient(qv, i)
+            ## Add contribution to fe
+            fe[i] += δu * dΩ
+            ## fe_shared[tx,i] += δu * dΩ
+            ## Loop over trial shape functions
+            for j in 1:n_basefuncs
+                ∇δu = shape_gradient(qv, j)
+                ## Add contribution to Ke
+                Ke[i, j] += (∇δu ⋅ ∇u) * dΩ
+            end
+        end
+    end
+    return
+end
+
+
+# gpu version of global assembly
+function assemble_gpu!(Kgpu, fgpu, cv, dh)
+    n_basefuncs = getnbasefunctions(cv)
+    assembler = start_assemble(Kgpu, fgpu; fillzero = false) ## has to be always false
```

Comment:
essentially yes because for every thread, it has its own version of the assembler referencing the same K so if it's true that means each thread will just erase what was done by the others so far. 

####  GPU Version
```julia
function Ferrite.start_assemble(K::CUSPARSE.CuSparseDeviceMatrixCSC{Tv, Ti}, f::CuDeviceVector{Tv}; fillzero::Bool = false) where {Tv, Ti}
    ## `fillzero` is not required in GPU version and should be always set to false, because different threads will
    ## instantiate their own instance of assembler, so it it's set to true, the results will be wrong.
    ## However, the interface is kept the same for consistency with the CPU version (in CPU multithreading we use the standard assembler).
    fillzero && (fillzero!(K); fillzero!(f))
    return GPUAssemblerSparsityPattern(K, f)
end
```

#### CPU Version
``` julia
function start_assemble(K::AbstractSparseMatrixCSC{T}, f::Vector = T[]; fillzero::Bool = true, maxcelldofs_hint::Int = 0) where {T}
    fillzero && (fillzero!(K); fillzero!(f))
    return CSCAssembler(K, f, zeros(Int, maxcelldofs_hint), zeros(Int, maxcelldofs_hint))
end
```
Honestly for the GPU assembly, this parameter `fillzero` is irrelevant whatsoever but I added it later when I started in CPU multithreading stuff, in which I use the already existing assembler object and by default `fillzero = true` which gives wrong answers on multithreading. so to make the gpu kernel looks like the cpu kernel, I added the same exact parameter in the gpu assembler.